### PR TITLE
feat(GraphQL): Support auth with custom DQL

### PR DESCRIPTION
**Pull request by minhaj-shakeel**
Fixes GRAPHQL-1171.
This PR adds support for @auth on custom DQL queries.
This PR also adds the parsing step to the DQL query.
The Algorithm tries to figure out the `queried type` at the root of the query either from the `root func` or `filters`.
Suppose if the root func is `func: type(Person)` or `func: has(Person.name)` or `func: eq(Person.age, 10)` then 
it infers that the queried type is `Person`. 
However, there are certain DQL queries on which @auth rules are not added currently since the queried type is not clear. 
for eg: 
```
me(func: uid(0x1, 0x2)) {
 ...
}
```
or the query:
```
me(func: has(name@en) {
 ...
}
```

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7775)

### CODE REVIEW

**Review by abhimanyusinghgaur — CHANGES_REQUESTED** [review]
> Reviewed 1 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7775)* status: 1 of 9 files reviewed, 8 unresolved discussions (waiting on @minhaj-shakeel and @pawanrawal)

___
*[graphql/dgraph/graphquery.go, line 217 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VnUdj8qebxsQ7UrXH:-M_VnUdk-n4K4NENu2Xr:bn33a3s) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/dgraph/graphquery.go#L217)):*
> ```
> writeFilterArguments(b, q.Func.Args, q.Func.Attr, q.Func.Name)
> ```
maybe just pass `q.Func` and use that in the function
___
*[graphql/dgraph/graphquery.go, line 223 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VnlEm71WOOQP7avLS:-M_VnlEm71WOOQP7avLT:b-4acqhk) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/dgraph/graphquery.go#L223)):*
> ```Go
> }
> 
> func maybeQuoteArg(fn string, arg interface{}) string {
> ```

I think we have a similar function in `resolve` pkg. Can't we reuse that here?
If it's a package dependency issue, we can move that function to a higher level package in the dependency tree (`schema` pkg), or in the package`x`.
___
*[graphql/e2e/auth/schema.graphql, line 991 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VpYbyDCXXS8-lAnrN:-M_VpYbyDCXXS8-lAnrO:b-lsqfyl) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/e2e/auth/schema.graphql#L991)):*
> ```
> $search: string
> ```
stale
___
*[graphql/e2e/auth/schema.graphql, line 997 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VqMgBBdAwvwG96z22:-M_VqMgBBdAwvwG96z23:buoa8hw) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/e2e/auth/schema.graphql#L997)):*
> ```
> orderdesc: val(age)
> ```
can this directly work here? as the map `age` is a map from owner uid to age value, while the root func here is iterating over the Issue uids. So, I think this might not produce the intended results after all.
This var query might be needed:
```
		iss as var(func: type(Issue)) @filter(has(Issue.owner)) {
			Issue.owner {
				age as User.age
			}
			ownerAge as max(val(age))
		}
```
Then use `val(ownerAge)` in the `orderdesc`.
___
*[graphql/resolve/custom_auth_query_test.yaml, line 26 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VxwAh7nt1wuNKK3Hi:-M_VxwAh7nt1wuNKK3Hj:b-lxoeh3) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/custom_auth_query_test.yaml#L26)):*
> ```YAML
>     }
> 
> - name : "custom DQL query with var block RBAC rules true"
> ```

can we also add another test where there are multiple var blocks in the DQL, and all of them are used in the user's query?
___
*[graphql/resolve/custom_auth_query_test.yaml, line 75 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VsRs92Kr6tlC7-1L3:-M_VsRs92Kr6tlC7-1L4:bng6oc1) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/custom_auth_query_test.yaml#L75)):*
> ```
> iss as var(func: uid(0x1)) @filter((uid(0x2) AND has(Issue.owner)))
> ```
:+1: 
why is the `AND has(Issue.owner)` here?
___
*[graphql/resolve/query_rewriter.go, line 641 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VvEPDC9M1YXwxuQD5:-M_VvEPE9nuQi7-BxMsx:boyhmlf) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/query_rewriter.go#L641)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
> 	args := query.Arguments()
> 	vars := make(map[string]string)
> 	for k, v := range args {
> 		// dgoapi.Request{}.Vars accepts only string values for variables,
> 		// so need to convert all variable values to string
> 		vStr, err := convertScalarToString(v)
> 		if err != nil {
> 			return nil, err
> 		}
> 		// the keys in dgoapi.Request{}.Vars are assumed to be prefixed with $
> 		vars["$"+k] = vStr
> 	}
> ```
</details>

we should put this in a function, this piece of code is being used at two places
___
*[graphql/resolve/query_rewriter.go, line 680 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VvZ_PBh_5jwaz-M6E:-M_VvZ_PBh_5jwaz-M6F:b-ksc1xm) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/query_rewriter.go#L680)):*
> ```
> 	typeName = extractTypeFromFilter(dgQuery.Filter)
> 	return typeName
> ```
`return extractTypeFromFilter(dgQuery.Filter)`

also, should we try to extract type from `order`?

**Review by minhaj-shakeel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7775)* status: 0 of 10 files reviewed, 8 unresolved discussions (waiting on @abhimanyusinghgaur and @pawanrawal)

___
*[graphql/dgraph/graphquery.go, line 217 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VnUdj8qebxsQ7UrXH:-M_xiMenE4NQv4DYNG5_:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/dgraph/graphquery.go#L217)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> writeFilterArguments(b, q.Func.Args, q.Func.Attr, q.Func.Name)
> ```
maybe just pass `q.Func` and use that in the function
</blockquote></details>

Done.
___
*[graphql/dgraph/graphquery.go, line 223 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VnlEm71WOOQP7avLS:-M_xiOcy9BFx2CYNUyh1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/dgraph/graphquery.go#L223)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

I think we have a similar function in `resolve` pkg. Can't we reuse that here?
If it's a package dependency issue, we can move that function to a higher level package in the dependency tree (`schema` pkg), or in the package`x`.
</blockquote></details>

Done.
___
*[graphql/e2e/auth/schema.graphql, line 991 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VpYbyDCXXS8-lAnrN:-M_xiQGz7gdsfqS_HWwr:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/e2e/auth/schema.graphql#L991)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> $search: string
> ```
stale
</blockquote></details>

Done.
___
*[graphql/e2e/auth/schema.graphql, line 997 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VqMgBBdAwvwG96z22:-M_xiRRN6qd6-Jcpqiy0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/e2e/auth/schema.graphql#L997)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> orderdesc: val(age)
> ```
can this directly work here? as the map `age` is a map from owner uid to age value, while the root func here is iterating over the Issue uids. So, I think this might not produce the intended results after all.
This var query might be needed:
```
		iss as var(func: type(Issue)) @filter(has(Issue.owner)) {
			Issue.owner {
				age as User.age
			}
			ownerAge as max(val(age))
		}
```
Then use `val(ownerAge)` in the `orderdesc`.
</blockquote></details>

Done.
___
*[graphql/resolve/custom_auth_query_test.yaml, line 26 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VxwAh7nt1wuNKK3Hi:-M_xiU2Q6F_r0TDfwtQQ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/custom_auth_query_test.yaml#L26)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

can we also add another test where there are multiple var blocks in the DQL, and all of them are used in the user's query?
</blockquote></details>

Done.
___
*[graphql/resolve/custom_auth_query_test.yaml, line 75 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VsRs92Kr6tlC7-1L3:-M_xiV340cYwNq2-Aonm:bzb3qs) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/custom_auth_query_test.yaml#L75)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> iss as var(func: uid(0x1)) @filter((uid(0x2) AND has(Issue.owner)))
> ```
:+1: 
why is the `AND has(Issue.owner)` here?
</blockquote></details>

because it is attached in the query.
___
*[graphql/resolve/query_rewriter.go, line 641 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VvEPDC9M1YXwxuQD5:-M_xibeZ3qKgAYpDb5UP:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/query_rewriter.go#L641)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> 	args := query.Arguments()
> 	vars := make(map[string]string)
> 	for k, v := range args {
> 		// dgoapi.Request{}.Vars accepts only string values for variables,
> 		// so need to convert all variable values to string
> 		vStr, err := convertScalarToString(v)
> 		if err != nil {
> 			return nil, err
> 		}
> 		// the keys in dgoapi.Request{}.Vars are assumed to be prefixed with $
> 		vars["$"+k] = vStr
> 	}
> ```
we should put this in a function, this piece of code is being used at two places
</blockquote></details>

Done.
___
*[graphql/resolve/query_rewriter.go, line 680 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-M_VvZ_PBh_5jwaz-M6E:-M_xidSJChwUBjZVZ0Pd:b-gr4804) ([raw file](https://github.com/dgraph-io/dgraph/blob/b54cc0cef8a9d9af158a1d47b085c7e13e0a4bae/graphql/resolve/query_rewriter.go#L680)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> 	typeName = extractTypeFromFilter(dgQuery.Filter)
> 	return typeName
> ```
`return extractTypeFromFilter(dgQuery.Filter)`

also, should we try to extract type from `order`?
</blockquote></details>

Done. Added an example of order also.

File: graphql/dgraph/graphquery.go [hunk 1]
```diff
@@ -112,7 +133,40 @@ func writeQuery(b *strings.Builder, query *gql.GraphQuery, prefix string) {
 	}
 }
 
-func writeUIDFunc(b *strings.Builder, uids []uint64, args []gql.Arg) {
+func writeNeedVar(b *strings.Builder, query *gql.GraphQuery) {
```
Comment (vmrajas) [inline graphql/dgraph/graphquery.go:136]: add a comment

File: graphql/dgraph/graphquery.go [hunk 2]
```diff
@@ -145,25 +206,35 @@ func writeRoot(b *strings.Builder, q *gql.GraphQuery) {
...
+		writeFilterArguments(b, q.Func)
 		x.Check2(b.WriteRune(')'))
 	}
 	writeOrderAndPage(b, q, true)
-	x.Check2(b.WriteRune(')'))
 }
 
-func writeFilterArguments(b *strings.Builder, args []gql.Arg) {
-	for i, arg := range args {
+func writeFilterArguments(b *strings.Builder, q *gql.Function) {
```
Comment (vmrajas) [inline graphql/dgraph/graphquery.go:224]: Add a comment. check if the condition needs to be updated.

File: graphql/e2e/auth/schema.graphql [hunk 3]
```diff
@@ -972,3 +972,112 @@ type Home @auth(
...
+}
+
+type IssueMap {
+	owner: ID!
+	count: Int
+}
+
+type Query {
+
+    queryIssueSortedByOwnerAge: [Issue] @custom(dql: """
```
Comment (vmrajas) [inline graphql/e2e/auth/schema.graphql:990]: Fix indentation if possible.

File: graphql/e2e/auth/auth_test.go [hunk 4]
```diff
@@ -322,6 +322,159 @@ func (s Student) add(t *testing.T) {
...
+			queryProjectsOrderByName{
+				name
+			}
+		}
+		`,
+			role:   "ADMIN",
+			result: `{"queryProjectsOrderByName":[{"name": "Project1"},{"name": "Project2"}]}`,
+		},
+		{
+			name: "RBAC OR filter query; RBAC fail",
```
Comment (vmrajas) [inline graphql/e2e/auth/auth_test.go:340]: nit: RBAC false or user1 projects.

File: graphql/e2e/auth/auth_test.go [hunk 5]
```diff
@@ -322,6 +322,159 @@ func (s Student) add(t *testing.T) {
 	require.JSONEq(t, result, string(gqlResponse.Data))
 }
 
+func TestAuthWithCustomDQL(t *testing.T) {
+	TestCases := []TestCase{
+		{
+			name: "RBAC OR filter query; RBAC Pass",
+			query: `
+		query{
```
Comment (vmrajas) [inline graphql/e2e/auth/auth_test.go:330]: nit: indent

File: graphql/resolve/custom_auth_query_test.yaml [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+          Column.inProject : Column.inProject {
+            Project.roles : Project.roles @filter(eq(Role.permission, "VIEW")) {
+              Role.assignedTo : Role.assignedTo @filter(eq(User.username, "user1"))
+            }
+          }
+        }
+      }
+    }
+
+- name : "Auth rules with deep filter missing JWT"
```
Comment (vmrajas) [inline graphql/resolve/custom_auth_query_test.yaml:154]: add an explanation field for easier understanding.

File: graphql/resolve/query_rewriter.go [hunk 7]
```diff
@@ -611,6 +630,176 @@ func rewriteAsGet(
...
+	if err != nil {
+		return nil, err
+	}
+
+	dqlReq := gql.Request{
+		Str:       dgQuery,
+		Variables: vars,
+	}
+	parsedResult, err := gql.Parse(dqlReq)
+	for _, qry := range parsedResult.Query {
```
Comment (vmrajas) [inline graphql/resolve/query_rewriter.go:648]: Add a comment

File: graphql/resolve/query_rewriter.go [hunk 8]
```diff
@@ -806,7 +995,7 @@ func (authRw *authRewriter) addAuthQueries(
 
 		// Adding the case of Query on interface in which None of the implementing type have
 		// Auth Query Rules, in that case, we also return simple query.
-		if typ.IsInterface() == true && implementingTypesHasAuthRules == false {
+		if typ.IsInterface() && !implementingTypesHasAuthRules {
```
Comment (vmrajas) [inline graphql/resolve/query_rewriter.go:998]: :+1:

File: graphql/resolve/query_rewriter.go [hunk 9]
```diff
@@ -1506,6 +1701,171 @@ func addSelectionSetFrom(
...
+			continue
+		}
+		if authRules := dqlHasAuthRules(fld, typ.Field(fldName).Type(), authRw); authRules {
+			return true
+		}
+	}
+	return false
+}
+
+func getFieldName(attr string) string {
```
Comment (vmrajas) [inline graphql/resolve/query_rewriter.go:1728]: May be add a TODO here or in query rewriter that dgraph directive with custom DQL auth won't work and is to be added.

**Review by vmrajas — APPROVED** [review]
LGTM

**Review by minhaj-shakeel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7775)* status: 0 of 10 files reviewed, 15 unresolved discussions (waiting on @abhimanyusinghgaur, @minhaj-shakeel, @pawanrawal, and @vmrajas)

___
*[graphql/e2e/auth/schema.graphql, line 990 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-MZmiFhxCIObd98LcGLw-r2-990:-MaSwm0G2Vv3IJxl0p3I:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/434cdcd6085462fb759f3171bcb787b5ce175fbb/graphql/e2e/auth/schema.graphql#L990)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

Fix indentation if possible.
</blockquote></details>

Done.
___
*[graphql/resolve/custom_auth_query_test.yaml, line 154 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-MZpZG4YRayolWeD0itY-r2-154:-MaSwnsE2pUTNeqZip7a:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/434cdcd6085462fb759f3171bcb787b5ce175fbb/graphql/resolve/custom_auth_query_test.yaml#L154)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

add an explanation field for easier understanding.
</blockquote></details>

Done.
___
*[graphql/resolve/query_rewriter.go, line 648 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-MZmiFhxCIObd98LcGLy-r2-648:-MaSwt0O19Mdj2akcxdV:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/434cdcd6085462fb759f3171bcb787b5ce175fbb/graphql/resolve/query_rewriter.go#L648)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

Add a comment
</blockquote></details>

Done.
___
*[graphql/resolve/query_rewriter.go, line 998 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-MZmiFhxCIObd98LcGLy-r2-998:-MaSwuLsD0xVQ-U_hjov:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/434cdcd6085462fb759f3171bcb787b5ce175fbb/graphql/resolve/query_rewriter.go#L998)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

:+1: 
</blockquote></details>

Done.
___
*[graphql/resolve/query_rewriter.go, line 1728 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7775#-MZmiFhxCIObd98LcGLy-r2-1728:-MaSxckH9ZrNskgwUHTb:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/434cdcd6085462fb759f3171bcb787b5ce175fbb/graphql/resolve/query_rewriter.go#L1728)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

May be add a TODO here or in query rewriter that dgraph directive with custom DQL auth won't work and is to be added.
</blockquote></details>

Done.

**Review by abhimanyusinghgaur — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7775)* status: 0 of 10 files reviewed, 7 unresolved discussions (waiting on @abhimanyusinghgaur, @minhaj-shakeel, @pawanrawal, and @vmrajas)

**Comment by minhaj-shakeel** [comment]
Fixes GRAPHQL-1171.
This PR adds support for @auth on custom DQL queries.
This PR also adds the parsing step to the DQL query.
The Algorithm tries to figure out the `queried type` at the root of the query either from the `root func` or `filters`.
Suppose if the root func is `func: type(Person)` or `func: has(Person.name)` or `func: eq(Person.age, 10)` then 
it infers that the queried type is `Person`. 
However, there are certain DQL queries on which @auth rules are not added currently since the queried type is not clear. 
for eg: 
```
me(func: uid(0x1, 0x2)) {
 ...
}
```
or the query:
```
me(func: has(name@en) {
 ...
}
```

**Comment by benwoodward** [comment]
Would be great if there was some example syntax. If I'm creating a `@custom` query, then I put it inside `type Query {}`. So presumably I add `@auth` directive rules to `type Query {}` as I would do if it were not a `@custom` query? Assuming this is the case, then if I have multiple `@custom` queries, can I create multiple `type Query {}` blocks, one for each `@custom` query, each block with its own `@auth` rules? How would you solve this problem?

**Pull request merged**